service/tablet_allocator.cc

```diff
@@ -12,12 +12,419 @@
+        // If target node is not balanced across shards, we will overload some shards.
+        // If tablets are not balanced in size, throughput will suffer because some shards will be idle sooner than others.
+        //
+        // FIXME: To handle the above, we should (1) rebalance the target node
+        // before migrating tablets from other nodes. If shards are balanced on the target node, the balancer
+        // will naturally distribute tablets to different shards. Also, (2) we should change this algorithm
+        // to be a generator for migrations and have a scheduler in the execution layer which pulls migrations
+        // from this algorithm, batches them and decides how many to execute.
+        //
+        // The scheduler decides in which order to execute the plan based on current activity in the system.
+        // We cannot just ask the planner for the next migration and stop when we hit overload on some shard,
+        // because that can lead to underutilization of the cluster. Just because the next migration is blocked
+        // by the target shard being busy doesn't mean we could not proceed with migrations for other shards
+        // which would be produced by the planner subsequently.
+
+        auto target_node = topo.find_node(target);
+        auto batch_size = target_node->get_shard_count();
+
+        // Compute per-shard load and candidate tablets.
+
+        for (auto&& [table, tmap] : _tm->tablets().all_tables()) {
+            std::optional<tablet_id> tid = tmap.first_tablet();
+            if (!tmap.transitions().empty()) {
+                // FIXME: The algorithm doesn't support balancing with active transitions yet. They must finish first.
+                lblogger.warn("Pending transitions active.");
+                co_return migration_plan();
+            }
+
+            for (const tablet_info& ti : tmap.tablets()) {
+                co_await coroutine::maybe_yield();
+                for (auto&& replica : ti.replicas) {
+                    if (!nodes.contains(replica.host)) {
+                        continue;
+                    }
+                    auto& node_load_info = nodes[replica.host];
+                    auto&& shard_load_info = node_load_info.shards[replica.shard];
+                    if (shard_load_info.tablet_count == 0) {
+                        node_load_info.candidates.push_back(replica.shard);
+                    }
+                    shard_load_info.tablet_count += 1;
+                    shard_load_info.candidates.emplace(global_tablet_id{table, *tid});
+                }
+                tid = tmap.next_tablet(*tid);
+            }
+        }
+
+        // Prepare candidate nodes and shards for heap-based balancing.
+
+        // heap which tracks most-loaded nodes in terms of avg_load.
+        std::vector<host_id> candidate_nodes;
+        candidate_nodes.reserve(nodes.size());
+        auto nodes_cmp = [&] (const host_id& a, const host_id& b) {
```

Comment:
That's not exactly the same spirit, since the container is provided externally. candidates_cmp() lives at the same scope as the container. So in that spirit, nodes_cmp should live on the stack like right now, where the container is.